crates/astria-sequencer/src/app.rs

```diff
@@ -418,6 +428,67 @@ impl App {
         (signed_txs, validated_txs)
     }
 
+    /// Executes the given transaction data, writing it to the app's `StateDelta`
+    /// and stores the results in `self.execution_result`.
+    ///
+    /// Will throw error if any transaction fails to decode or execute as these
+    /// transactions should've been filtered in prepare proposal.
+    ///
+    /// Returns the transactions which were [`SignedTransaction`] form.
+    #[instrument(name = "App::execute_block_data", skip_all, fields(
+        tx_count = txs.len()
+    ))]
+    async fn execute_block_data(
+        &mut self,
+        txs: VecDeque<bytes::Bytes>,
+    ) -> Result<Vec<SignedTransaction>, anyhow::Error> {
+        let mut signed_txs = Vec::new();
+        for tx in txs {
+            let signed_tx = signed_transaction_from_bytes(&tx)
+                .map_err(|_| anyhow!("failed to decode tx bytes to signed transaction"))?;
+            match self.execute_tx(signed_tx).await {
+                Ok(signed_tx) => {
+                    signed_txs.push(signed_tx);
+                }
+                Err(_e) => {
+                    // all txs in the proposal should be deserializable and executable,
+                    // if any txs were not deserializeable or executable, they would not have been
+                    // returned by `execute_block_data`, something is wrong
+                    bail!("all transactions should succeed at this stage");
+                }
+            }
+        }
+        Ok(signed_txs)
+    }
+
+    /// Executes the given transaction data, writing it to the app's `StateDelta`
+    /// and stores the results in `self.execution_result`.
+    ///
+    /// Returns the signed transaction if it successfully executed else an error.
+    #[instrument(name = "App::execute_tx", skip_all)]
+    async fn execute_tx(
+        &mut self,
+        signed_tx: SignedTransaction,
+    ) -> Result<SignedTransaction, anyhow::Error> {
+        // store transaction execution result, indexed by tx hash
+        let tx_hash = Sha256::digest(signed_tx.to_raw().encode_to_vec());
+        match self.deliver_tx(signed_tx.clone()).await {
+            Ok(events) => {
+                self.execution_result.insert(tx_hash.into(), Ok(events));
+                return Ok(signed_tx);
+            }
+            Err(e) => {
+                debug!(
+                    transaction_hash = %telemetry::display::hex(&Sha256::digest(tx_hash)),
+                    error = AsRef::<dyn std::error::Error>::as_ref(&e),
+                    "failed to execute transaction"
+                );
+                self.execution_result.insert(tx_hash.into(), Err(e));
```

Comment:
I couldn't figure out how to copy the error so I couldn't bubble it up. I don't like returning this error message but I wasn't sure what a better thing to do was.